pkg/pool-weighted/contracts/managed/ManagedPool.sol

```diff
@@ -587,6 +710,73 @@ contract ManagedPool is ManagedPoolSettings {
         return ComposablePoolLib.dropBpt(registeredTokens, registeredBalances);
     }
 
+    // Circuit Breakers
+
+    /**
+     * @dev Get the current BPT price, given its components (e.g., the end state of an operation).
+     */
+    function _getBptPrice(
+        uint256 virtualSupply,
+        uint256 normalizedWeight,
+        uint256 upscaledBalance
+    ) private pure returns (uint256) {
+        return virtualSupply.mulUp(normalizedWeight).divDown(upscaledBalance);
+    }
+
+    /**
+     * @dev Trip the circuit breaker if the current BPT price is less than the lower bound, or greater than the
+     * upper bound. For performance reasons, check for a zero bound (= no circuit breaker set) externally,
+     * avoiding computation of the current BPT price when no breaker is set.
+     */
+    function _checkCircuitBreaker(
+        uint256 bptPriceBound,
+        uint256 virtualSupply,
+        uint256 normalizedWeight,
+        uint256 upscaledBalance,
+        bool isLowerBound
+    ) private pure {
+        uint256 currentBptPrice = _getBptPrice(virtualSupply, normalizedWeight, upscaledBalance);
+
+        _require(
+            isLowerBound ? currentBptPrice >= bptPriceBound : currentBptPrice <= bptPriceBound,
+            Errors.CIRCUIT_BREAKER_TRIPPED
+        );
+    }
+
+    /**
+     * @dev Check circuit breakers for a set of tokens. The given virtual supply is what it will be post-operation:
+     * this includes any pending external fees, and the amount of BPT exchanged (swapped, minted, or burned) in the
+     * current operation.
+     *
+     * First get the appropriate bound (lowerBound for joins, since increasing the token balance decreases the BPT
+     * price; upperBound for exits), short-circuiting if it is zero. Adjust the balance by the poolDelta amount,
+     * and call `_checkCircuitBreaker` to test the current token.
+     */
+    function _checkCircuitBreakers(
+        uint256 virtualSupply,
+        PoolDelta memory poolDelta,
+        bool isJoin
+    ) private view {
+        for (uint256 i = 0; i < poolDelta.tokens.length; i++) {
+            // If it is a join, tokens are coming in, and we're checking the lower bound, so isLowerBound = isJoin.
+            uint256 bptPriceBound = _getCurrentCircuitBreakerBound(
+                poolDelta.tokens[i],
+                poolDelta.normalizedWeights[i].complement(),
+                isJoin
+            );
```

Comment:
I don't think this is true. It's true that in token swaps we can reason about how the BPT price of different tokens will change, we can't in situations where the total supply is changing.

e.g. Consider a 50:50 pool with tokens A and B. Say token A has a tight upper circuit breaker so if the price increases a little bit then the breaker will trip. If I do a huge join with token B then the join will succeed (assuming I don't trip a breaker on token B) despite the fact that I've massively inflated the BPT price of token A